C99 compatibility fixes for build-time feature probing.

Include <unistd.h> for the close function.  Do not call the
undeclared __tls_get_addr function; rely on __thread support
for feature probing.  Define _GNU_SOURCE to obtain prototypes
of fallocate and memmem.

Submitted upstream: <https://github.com/fluent/fluent-bit/pull/6704>

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 90d32da30..c6315bb2a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1012,8 +1012,7 @@ if(NOT FLB_POSIX_TLS)
   check_c_source_compiles("
    __thread int a;
    int main() {
-       __tls_get_addr(0);
-       return 0;
+       return a;
    }" FLB_HAVE_C_TLS)
   if(FLB_HAVE_C_TLS)
     FLB_DEFINITION(FLB_HAVE_C_TLS)
diff --git a/lib/chunkio/CMakeLists.txt b/lib/chunkio/CMakeLists.txt
index 233bce35c..9edcc2b84 100644
--- a/lib/chunkio/CMakeLists.txt
+++ b/lib/chunkio/CMakeLists.txt
@@ -81,6 +81,7 @@ endif()
 
 # fallocate(2) support
 check_c_source_compiles("
+  #define _GNU_SOURCE
   #include <fcntl.h>
   int main() {
      fallocate(0,0,0);
diff --git a/lib/monkey/CMakeLists.txt b/lib/monkey/CMakeLists.txt
index 198251d8e..bd3eb8b4d 100644
--- a/lib/monkey/CMakeLists.txt
+++ b/lib/monkey/CMakeLists.txt
@@ -181,8 +181,7 @@ if(NOT MK_PTHREAD_TLS)
   check_c_source_compiles("
      __thread int a;
      int main() {
-         __tls_get_addr(0);
-         return 0;
+         return a;
      }" HAVE_C_TLS)
 
   if(HAVE_C_TLS)
diff --git a/lib/monkey/mk_core/CMakeLists.txt b/lib/monkey/mk_core/CMakeLists.txt
index 86cc7fcb4..f71d6addc 100644
--- a/lib/monkey/mk_core/CMakeLists.txt
+++ b/lib/monkey/mk_core/CMakeLists.txt
@@ -61,6 +61,7 @@ set(src "${src}"
   )
 
 check_c_source_compiles("
+  #define _GNU_SOURCE
   #include <string.h>
   int main() {
      char  haystack[] = \"1234\";
@@ -145,7 +146,7 @@ endif()
 
 configure_file(
   "${CMAKE_CURRENT_SOURCE_DIR}/../include/monkey/mk_core/mk_core_info.h.in"
-  "${PROJECT_BINARY_DIR}/include/monkey/mk_core/mk_core_info.h" 
+  "${PROJECT_BINARY_DIR}/include/monkey/mk_core/mk_core_info.h"
   )
 
 add_library(mk_core STATIC ${src})
